### ISSUE TITLE
bgzip reindex performance

### ISSUE THREAD

**Issue by jacrayner** [samtools/htslib#378]
Hi,

I have noticed that bgzip's reindex functionally seem slower that is necessary. 

The index in made up of the uncompressed and compressed offsets for each bgzip block in a file possibly excluding the final block. I think all this information is contained within the bgzip header and footer i.e. the BSIZE feild and the ISIZE field (page 11: https://samtools.github.io/hts-specs/SAMv1.pdf) 

However the current implementation is passing the entire file block-wize to  bgzf_read() in bgzf.c which I think is decompressing the file to rebuild the index.

Is there any reason why the reindexing functionality can be implemented by just seeking through the file and just pulling out the BSIZE and ISIZE fields?

By the way I am interested in reindex performance due to  another issued I raised in parallel to this: 
https://github.com/samtools/htslib/issues/377

**Comment by jkbonfield** [comment]
Agreed decompressing the entire data stream seems redundant given the uncompressed size is known upfront.  (This is also one reason why samtools index on a CRAM file is much faster than a BAM file, as CRAM operates in the manner you describe.)

I suspect the reason it is done this way is simply that is easier to code; maybe zlib lacks the appropriate API to decode only the gzip header and not the contents?  However it's very low priority for us as we have many other things more significant (to us).

Are you proposing implementing this and submitting a pull request?

**Comment by jacrayner** [comment]
yes that sound correct, it was not obvious to me how to achieve this within the framework of the bgzf API and probably reindex is not a much used feature. Yes thanks as someone more familiar with the code base can see no obvious reason why this shouldn't or can't be done then I will see what I can do.

**Comment by jrandall** [comment]
@jacrayner seems like a good fix - should we continue to expect an eventual PR for this?

**Comment by jacrayner** [comment]
Hi @jrandall, it is something I am looking at now but there seems to be lots of more significant work on bgzip and I am tempted to wait for @jkbonfield to merge some of his PRs. Also #402 may effect this if bgzf stops manually striping the gzip header/footer.
